### DISCUSSION TITLE
How to get value of a Variable of type int conserving its type

### DISCUSSION THREAD

**Discussion by waltergallegog** [lava-nc/lava#261] [Q&A] [answered]
I'm trying to access the value of a `Var` of type `int`, using the `get()` method. However, the type I'm getting is `float`.
Is this expected?
How can I get the value without it being converted to float?.

Here is the relevant snips of my code. The var I want to access is `in_data`

```python
# Minimal process with an InPort
class P2(AbstractProcess):
	def __init__(self, **kwargs):
		super().__init__(**kwargs)
		self.inp = InPort(shape=(1,))
		self.in_data = Var(shape=(1,), init=0)


# A minimal PyProcModel implementing P2
@implements(proc=P2, protocol=LoihiProtocol)
@requires(CPU)
@tag('floating_pt')
class PyProcModelB(PyLoihiProcessModel):
	inp    : PyInPort = LavaPyType(PyInPort.VEC_DENSE, int)
	in_data: int      = LavaPyType(int, int)
	def run_spk(self):
		self.in_data = self.inp.recv()
		dummy = self.in_data[0]
		print(f"Received input data in P2: {dummy}")
		print(f"Received input type in P2: {type(dummy)}")

# main
sender = P1()
recv   = P2()
sender.run(RunSteps(num_steps=1), Loihi1SimCfg())
recv_data = recv.in_data.get()[0]

```

Here you can get the full script:
https://github.com/waltergallegog/code-sharing/blob/master/lava/discussions/get_int.py

When running it I obtain `int64` type inside the process, but `float64` from the main:

```
Received input data in P2: 5
Received input type in P2: <class 'numpy.int64'>
Received input data in main: 5.0
Received input type in main: <class 'numpy.float64'>

```

**Comment by tim-shea** [comment]
I think this is a bug, though I'm not 100%. When I look into the var access code in abstract process model, it doesn't seem there is any mechanism to propagate the dtype of a numpy array from the process model to the process, which is what you ultimately access from main. @mgkwill do you have any insight into this one?

**Comment by mgkwill** [comment, marked as answer]
@waltergallegog have your tried the same code replacing `@tag('floating_pt')` with `@tag('fixed_pt')`?

**Reply by mgkwill** [reply]
I tried it out with `@tag('fixed_pt')` and still the same result. 
```
Sent output data of P1: [5]
Received input data in P2: 5
Received input type in P2: <class 'numpy.int64'>
Received input type in repr: 5
Received input data in main: 5.0
Received input type in main: <class 'numpy.float64'>
Received input type in main repr: Variable: in_data
    shape: (1,)
    init: 0
    shareable: True
    value: [5.]
```

I looked through the code and it appears that when setting a var, we send as np.float64 automatically, but there is not a provision for Int: 

https://github.com/lava-nc/lava/blob/7ad8e488c1859d92c15833689c3dbf978166d9ef/src/lava/magma/runtime/runtime.py#L430-L435

It looks like this change was added during enablement of float for get/set var in https://github.com/lava-nc/lava/pull/69 by @PhilippPlank at https://github.com/lava-nc/lava/pull/69/files#diff-c65c489aca051f74137b2f1f1e8c67e393132ddb0fefa29aafc5d4cec47b7ceeL325

As I think @tim-shea suggested we should file a bug and implement some sort of type metadata communication in the send and receive to preserve types.

**Comment by mgkwill** [comment]
I created a bug https://github.com/lava-nc/lava/issues/283

**Comment by mgkwill** [comment]
@waltergallegog Thanks for pointing this out and for providing a code example!

**Reply by waltergallegog** [reply]
@mgkwill Thanks for the support and for the test. Sorry I could not do it but due to timezone differences I just saw your request.
Now that there is a bug created I have marked as accepted answer the one with the explanation of the issue.
